File: leetcode/python/P0445_AddTwoNumbers2.py
# https://leetcode.com/problems/add-two-numbers-ii/

from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val: int = 0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        def reverseList(head: Optional[ListNode]) -> Optional[ListNode]:
            if head is None or head.next is None:
                return head
            rest = reverseList(head.next)
            head.next.next = head

            head.next = None

            return rest

        if l1 is None:
            return l2
        if l2 is None:
            return l1

        l1 = reverseList(l1)
        l2 = reverseList(l2)

        h1, h2 = l1, l2
        h3, h3_head = None, None

        carry = 0
        while h1 is not None or h2 is not None:
            val = carry
            if h1 is not None:
                val += h1.val
                h1 = h1.next

            if h2 is not None:
                val += h2.val
                h2 = h2.next

            node = ListNode(val % 10)
            carry = val // 10

            if h3 is None:
                h3 = node
                h3_head = node
            else:
                h3.next = node
                h3 = h3.next

        if carry == 1:
            # h3 can't be None here, having carry means we have
            # at least one element
            assert h3 is not None
            h3.next = ListNode(1)

        h3 = reverseList(h3_head)
        return h3

File: leetcode/cpp/P0445_AddTwoNumbers2.cpp
#include <stack>

/**
 * Definition for singly-linked list.
 */
struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

// Note: different approach from Python version.
// This one uses stacks to store list elements in reversed order.
// Time: O(n); Space: O(n)
class Solution {
 public:
  ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    std::stack<int> st1;
    std::stack<int> st2;

    ListNode* h1 = l1;
    while (h1) {
      st1.push(h1->val);
      h1 = h1->next;
    }

    ListNode* h2 = l2;
    while (h2) {
      st2.push(h2->val);
      h2 = h2->next;
    }

    int max_elems = std::max(st1.size(), st2.size());
    ListNode* result_head = nullptr;
    int carry = 0;
    for (int ii = 0; ii < max_elems; ++ii) {
      int el1 = 0;
      if (!st1.empty()) {
        el1 = st1.top();
        st1.pop();
      }

      int el2 = 0;
      if (!st2.empty()) {
        el2 = st2.top();
        st2.pop();
      }

      int crt = el1 + el2 + carry;
      carry = crt / 10;
      crt %= 10;

      ListNode* newhead = new ListNode(crt);
      newhead->next = result_head;
      result_head = newhead;
    }

    if (carry == 1) {
      ListNode* newhead = new ListNode(1);
      newhead->next = result_head;
      result_head = newhead;
    }

    return result_head;
  }
};